src/Firestore/Query.php

```diff
@@ -560,47 +598,166 @@ private function buildPosition(array $fieldValues, $before)
                         $c,
                         $this->childPath($collection, $value)
                     );
-                } elseif ($value instanceof DocumentReference) {
+                } elseif ($value instanceof DocumentReference || $value instanceof DocumentSnapshot) {
                     $name = $value->name();
                     if (!$this->isPrefixOf($collection, $name)) {
-                        throw new \BadMethodCallException(sprintf(
+                        throw new \InvalidArgumentException(sprintf(
                             '%s is not a part of the query result set and ' .
                             'cannot be used as a query boundary',
                             $name
                         ));
                     }
                 } else {
-                    throw new \BadMethodCallException(
+                    throw new \InvalidArgumentException(
                         'The corresponding value for DOCUMENT_ID must be a ' .
                         'string or a DocumentReference.'
                     );
                 }
 
-                if ($value->parent()->name() !== $collection) {
-                    throw new \BadMethodCallException(
+                $parent = ($value instanceof DocumentReference)
+                    ? $value->parent()->name()
+                    : $value->reference()->parent()->name();
+
+                if ($parent !== $collection) {
+                    throw new \InvalidArgumentException(
                         'Only direct children may be used as query boundaries.'
                     );
                 }
             }
         }
 
+        return $this->newQuery([
+            $key => [
+                'before' => $before,
+                'values' => $this->valueMapper->encodeValues($fieldValues)
+            ],
+            'orderBy' => $orderBy
+        ], true);
+    }
+
+    /**
+     * Build cursors for document snapshots.
+     *
+     * @param DocumentSnapshot $snapshot The document snapshot
+     * @return array[] A list, where position 0 is fieldValues and position 1 is orderBy.
+     */
+    private function snapshotPosition(DocumentSnapshot $snapshot, array $orderBy)
+    {
+        $append = true;
+        $direction = self::DIR_ASCENDING;
+
+        $fieldValues = [];
+
+        // If the list of orderBy clauses already contains __name__, use it unchanged.
+        $append = !(bool) array_filter($orderBy, function ($order) {
```

Comment:
s/append/appendName/ ?

And then please just have an all-encompassing `if(appendName) { ... }` that includes both the `where` and the `orderBy` block below.